### PR TITLE
initial drop — for review

### PR DESCRIPTION
**Pull request by costa**

### CODE REVIEW

File: ffmprb.gemspec [hunk 1]
```diff
@@ -4,28 +4,27 @@ $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
...
-  else
-    raise "RubyGems 2.0 or newer is required to protect against public gem pushes."
-  end
+  spec.summary       = "ffmprb is your audio/video montage friend, based on https://ffmpeg.org"
+  spec.description   = "A DSL (Damn-Simple Language) and a micro-engine for ffmpeg and ffriends"
+  spec.homepage      = 'https://github.com/showbox-oss/ffmprb'
 
   spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
-  spec.bindir        = "exe"
+  spec.bindir        = 'exe'
```
Comment (tanin) [inline ffmprb.gemspec:17]: There is no such dir, do we still need this setup here?

File: lib/ffmprb.rb [hunk 2]
```diff
@@ -1,5 +1,83 @@
...
+  Util::IoBuffer.blocks_max = 1024
+  Util::IoBuffer.block_size = 64*1024
+  Util::IoBuffer.timeout = 9
+
+  class << self
+
+    def process(*args, &blk)
+      logger.debug "Starting process with #{args} in #{blk.source_location}"
+      Process.new.tap do |process|
+        if blk
```
Comment (tanin) [inline lib/ffmprb.rb]: return unless blk

File: lib/ffmprb.rb [hunk 3]
```diff
@@ -1,5 +1,83 @@
...
+        logger.level = debug ? Logger::DEBUG : Logger::INFO
+      end
+    end
+
+    def logger=(logger)
+      @logger.close  if @logger
+      @logger = logger
+    end
+
+    def find_silence(input_file, output_file)
```
Comment (tanin) [inline lib/ffmprb.rb]: Move to separate module/class

File: lib/ffmprb/file.rb [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+          )
+        thr = Util::Thread.new do
+          buff.flush!
+          Ffmprb.logger.debug "IoBuffering from #{input_fifo_file.path} to #{output_fifo_file.path} ended"
+          input_fifo_file.remove
+          output_fifo_file.remove
+        end
+        Ffmprb.logger.debug "IoBuffering from #{input_fifo_file.path} to #{output_fifo_file.path} started"
+
+        yield buff  if block_given?  # XXX a hidden option
```
Comment (tanin) [inline lib/ffmprb/file.rb]: WTF?

File: lib/ffmprb/file.rb [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+          ->{
+            Ffmprb.logger.debug "Trying to open #{input_fifo_file.path} for reading+buffering"
+            ::File.open(input_fifo_file.path, 'r')
+          },
+          ->{
+            Ffmprb.logger.debug "Trying to open #{output_fifo_file.path} for buffering+writing"
+            ::File.open(output_fifo_file.path, 'w')
+            }
+          )
+        thr = Util::Thread.new do
```
Comment (tanin) [inline lib/ffmprb/file.rb:19]: need explanation

File: lib/ffmprb/file.rb [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+      end
+
+      def create(path)
+        new(path: path, mode: :write).tap do |file|
+          Ffmprb.logger.debug "Created file with path: #{file.path}"
+        end
+      end
+
+      def open(path)
+        new(path: (path.respond_to?(:path)? path.path : path), mode: :read).tap do |file|
```
Comment (tanin) [inline lib/ffmprb/file.rb:39]: path can be File? So what for it can be opened if it already a file object?

File: lib/ffmprb/file.rb [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+    def extname
+      ::File.extname path
+    end
+
+    def length
+      @duration ||= probe['format']['duration']
+      return @duration.to_f  if @duration
+
+      @duration = probe(true)['frames'].reduce(0){|sum, frame| sum + frame['pkt_duration_time'].to_f}
```
Comment (tanin) [inline lib/ffmprb/file.rb]: comment pls

File: lib/ffmprb/file.rb [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+      video: true,
+      audio: nil
+    )
+      audio = File.temp('.mp3')  if audio == true
+      video = File.temp('.jpg')  if video == true
+
+      Ffmprb.logger.debug "Snap shooting files, video path: #{video ? video.path : 'NONE'}, audio path: #{audio ? audio.path : 'NONE'}"
+
+      raise Error, "Incorrect output extname (must be .jpg)"  unless !video || video.extname =~ /jpe?g$/
+      raise Error, "Incorrect audio extname (must be .mp3)"  unless !audio || audio.extname =~ /mp3$/
```
Comment (tanin) [inline lib/ffmprb/file.rb]: Did you mean this?
if video && !(video.extname =~ /jpe?g$/)

File: lib/ffmprb/file.rb [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+      raise Error, "Can sample either video OR audio UNLESS a block is given"  unless block_given? || (!!audio != !!video)
```
Comment (tanin) [inline lib/ffmprb/file.rb]: unclear condition :)

File: lib/ffmprb/file.rb [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+      cmd = " -i #{path}"
+      cmd << " -deinterlace -an -ss #{at} -r 1 -vcodec mjpeg -f mjpeg #{video.path}"  if video
+      cmd << " -vn -ss #{at} -t 1 -f mp3 #{audio.path}"  if audio
+      Util.ffmpeg cmd
+
+      return video || audio  unless block_given?
```
Comment (tanin) [inline lib/ffmprb/file.rb:162]: what if both video and audio exist?

File: lib/ffmprb/filter.rb [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+      def afade_in(duration=1, input=nil, output=nil)
+        inout "afade=in:d=#{duration}", input, output
+      end
+
+      def afade_out(duration=1, input=nil, output=nil)
+        inout "afade=out:d=#{duration}", input, output
+      end
+
+      # NOTE inputs order matters
+      def amix(inputs, output=nil)
```
Comment (tanin) [inline lib/ffmprb/filter.rb]: inputs better should be a hash with an exact index number(key)

File: lib/ffmprb/filter.rb [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+      end
+
+      def black_source(duration, resolution=nil, fps=nil, output=nil)
+        filter = "color=black:d=#{duration}"
+        filter << ":s=#{resolution}"  if resolution
+        filter << ":r=#{fps}"  if fps
+        inout filter, nil, output
+      end
+
+      def fade_out_alpha(duration=1, input=nil, output=nil)
```
Comment (tanin) [inline lib/ffmprb/filter.rb:48]: should be two separate methods fade_out and alpha, isn't is?

File: lib/ffmprb/filter.rb [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        if crop[:bottom] > 0 && crop[:top]
+          raise Error, "Must specify two of {top, bottom, height} at most"  if crop[:height]
+          crop[:height] = 1 - crop[:bottom] - crop[:top]
+        elsif crop[:height] > 0
+          if !crop[:top] && crop[:bottom] > 0
+            crop[:top] = 1 - crop[:height] - crop[:bottom]
+            exps << "y=in_h*#{crop[:top]}"
+          end
+        end
+        exps << "h=in_h*#{crop[:height]}"
```
Comment (tanin) [inline lib/ffmprb/filter.rb:107]: Both w and h calculation could be dried out by moving the generic logic to a separate method.  It looks like the same code with different params.

File: lib/ffmprb/filter.rb [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+      def scale_pad_fps(width, height, fps, input=nil, output=nil)
+        inout [
+          *scale(width, height),
+          *pad(width, height),
+          *fps(fps)
+        ].join(', '), input, output
+      end
+
+      def silencedetect(input=nil, output=nil)
+        inout "silencedetect=d=2:n=-30dB", input, output
```
Comment (tanin) [inline lib/ffmprb/filter.rb]: -30dB should be constant оr config

File: lib/ffmprb/filter.rb [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+      def trim(st, en, input=nil, output=nil)
+        inout "trim=#{[st, en].compact.join ':'}, setpts=PTS-STARTPTS", input, output
+      end
+
+      def volume(volume, input=nil, output=nil)
+        inout "volume='#{volume_exp volume}':eval=frame", input, output
+      end
+
+      def volume_exp(volume)  # NOTE volume is an ordered Hash
+        return volume  unless volume.respond_to?(:each)
```
Comment (tanin) [inline lib/ffmprb/filter.rb]: volume.is_a? Hash

File: lib/ffmprb/process.rb [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+            )  if silent.end_at
+          end
+          overlay in_over.volume ducked_overlay_volume
+          Ffmprb.logger.debug "Ducking audio with volumes: {#{ducked_overlay_volume.map{|t,v| "#{t}: #{v}"}.join ', '}}"
+        end
+
+      end
+    end
+
+    def initialize(*args, &blk)
```
Comment (tanin) [inline lib/ffmprb/process.rb:54]: This args and blk are not in use here. Why?

File: lib/ffmprb/process/input.rb [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+      class Cropped < Input
+
+        attr_reader :crop_ratios
+
+        def initialize(unfiltered, crop:)
+          @io = unfiltered
+          self.crop_ratios = crop
+        end
+
+        def filters_for(lbl, process:, video: true, audio: true)
```
Comment (tanin) [inline lib/ffmprb/process/input.rb]: what is lbl?

File: lib/ffmprb/process/output.rb [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+
+      def initialize(io, only: nil, resolution: Ffmprb::QVGA, fps: 30)
+        @io = io
+        @channels = [*only]
+        @channels = nil  if @channels.empty?
+        @resolution = resolution
+        @fps = 30
+      end
+
+      def options(process)
```
Comment (tanin) [inline lib/ffmprb/process/output.rb:17]: too long function

File: lib/ffmprb/process/output.rb [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+        # XXX TODO manage stream labels through process
+        raise Error, "Nothing to roll..."  if @reels.select(&:reel).empty?
+        raise Error, "Supporting just full_screen for now, sorry."  unless @reels.all?(&:full_screen?)
+
+        filters = []
+
+        # Concatting
+        segments = []
+        Ffmprb.logger.debug "Concatting segments: start"
```
Comment (tanin) [inline lib/ffmprb/process/output.rb]: change the message to something like: "preparing option"

File: lib/ffmprb.rb [hunk 20]
```diff
@@ -1,5 +1,89 @@
...
+      logger.debug "Finding silence (#{input_file.path}->#{output_file.path})"
+      filters = Filter.silencedetect
+      options = " -i #{input_file.path}#{Filter.complex_options filters} #{output_file.path}"
+      silence = []
+      Util.ffmpeg(options).split("\n").each do |line|
+        next  unless line =~ /^\[silencedetect\s.*\]\s*silence_(\w+):\s*(\d+\.?d*)/
```
Comment (tanin) [inline lib/ffmprb.rb]: silencedetect is not enough here. Should use ebur128 to detect it as well as for loudness normalization.

File: lib/ffmprb/file.rb (hunk 6 above)
Comment (costa) [inline lib/ffmprb/file.rb:39]: This is `Ffmprb::File`, we set the rules — for convenience: here we can create a `Ffmprb::File` off a path string, or a ruby `File`, or even another `Ffmprb::File`. It's convenient. If usecases aren't clear, I'll elaborate.

File: lib/ffmprb/filter.rb (hunk 12 above)
Comment (costa) [inline lib/ffmprb/filter.rb:48]: no, it's fade out alpha (alpha is to be faded out)

File: lib/ffmprb/filter.rb (hunk 13 above)
Comment (costa) [inline lib/ffmprb/filter.rb:107]: 2 < 3

**Comment by costa** [comment]
@tanin I'm done with the code review, say OK if I may close the branch.

**Pull request closed without merging**